pkg/standalone-utils/contracts/relayer/EulerWrapping.sol

```diff
@@ -0,0 +1,126 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-interfaces/contracts/standalone-utils/IEulerToken.sol";
+import "@balancer-labs/v2-interfaces/contracts/vault/IVault.sol";
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/Address.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/SafeERC20.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/math/FixedPoint.sol";
+
+import "@balancer-labs/v2-pool-utils/contracts/lib/ExternalCallLib.sol";
+
+import "./IBaseRelayerLibrary.sol";
+
+/**
+ * @title EulerWrapping
+ * @notice Allows users to wrap and unwrap Euler tokens
+ * @dev All functions must be payable so they can be called from a multicall involving ETH
+ */
+abstract contract EulerWrapping is IBaseRelayerLibrary {
+    using Address for address payable;
+    using SafeERC20 for IERC20;
+    using FixedPoint for uint256;
+
+    address private immutable _eulerProtocol;
+
+    /**
+     * @dev Euler protocol needs to be approved to transer mainToken
+     * @param eulerProtocol - the address of the euler protocol
+     */
+    constructor(address eulerProtocol) {
+        _eulerProtocol = eulerProtocol;
+    }
+
+    function wrapEuler(
+        IEulerToken wrappedToken,
+        address sender,
+        address recipient,
+        uint256 amount,
+        uint256 outputReference
+    ) external payable {
+        if (_isChainedReference(amount)) {
+            amount = _getChainedReferenceValue(amount);
+        }
+
+        IERC20 underlying = IERC20(wrappedToken.underlyingAsset());
+
+        // The wrap caller is the implicit sender of tokens, so if the goal is for the tokens
+        // to be sourced from outside the relayer, we must first pull them here.
+        if (sender != address(this)) {
+            require(sender == msg.sender, "Incorrect sender");
+            _pullToken(sender, underlying, amount);
+        }
+
+        underlying.safeApprove(_eulerProtocol, amount);
+
+        // Deposit MainToken into EulerToken
+        // 0 for the Euler primary account
+        wrappedToken.deposit(0, amount);
+
+        uint256 withdrawnWrappedAmount = wrappedToken.balanceOf(address(this));
+
+        if (recipient != address(this)) {
+            IERC20(wrappedToken).safeApprove(address(this), withdrawnWrappedAmount);
+            IERC20(wrappedToken).safeTransferFrom(address(this), recipient, withdrawnWrappedAmount);
```

Comment:
For sending tokens out of this contract, we should use `transfer` instead of `transferFrom` (no approval required).
```suggestion
            IERC20(wrappedToken).safeTransfer(recipient, withdrawnWrappedAmount);
```